### ISSUE TITLE
Serenity Screenplay samples of Waits

### ISSUE THREAD

**Issue by SiKing** [serenity-bdd/serenity-core#2064]
Just started coding my first test(s) in Serinity Screenplay pattern. I am struggling to find examples of waits. Most examples use demo webpages that never never transition to another page, so there is no need for synchronization / waits. Further, my IDE (Eclipse) does not automatically do static imports, and most examples do not show the import statements.

I am trying to model the following scenario:
1. launch our app
2. enter username, press enter
3. the page reloads, and **only now** can you see the password field
4. enter password, press enter
5. app loads

Note that our lower environments are quite slow, so there is a need to wait almost every step.

Here is what I came up with:
```
import static org.hamcrest.Matchers.is;  // deprecated?


@Override
public <T extends Actor> void performAs(T actor) {
	actor.wasAbleTo(Open.browserOn().the(applicationLandingPage));
	actor.attemptsTo(Enter.theValue("...").into(LoginForm.USERNAME).thenHit(Keys.ENTER));
	Wait.until(Enabled.of(LoginForm.PASSWORD).viewedBy(actor).asAQuestion(), is(true));
	actor.attemptsTo(Enter.theValue("...").into(LoginForm.PASSWORD).thenHit(Keys.ENTER));
}
```

My `Wait` here seems quite awkward, and most of the time does not work. I end up with "stale element reference: element is not attached to the page document". If I change it to a `Thread.sleep(10000)` then it works, most of the time.

Could you suggest a better approach. Also, pointer to some other examples of different waits would be appreciated.

**Comment by globalworming** [comment]
There are multiple approaches how to do waits, see http://thucydides.info/docs/serenity-staging/#_working_with_timeouts for examples.
If your page is performing badly in general, you can increase implicit wait globally by setting `webdriver.timeouts.implicitlywait`, see http://thucydides.info/docs/serenity-staging/#_serenity_system_properties_and_configuration


Screenplay specific you already provides an example. More things you can do:
```
// perform Wait
actor.attemptsTo(Wait.until(
    LoginForm.PASSWORD, WebElementStateMatchers.isEnabled()
).forNoLongerThan(30).seconds())

// assert with waiting
actor.should(EventualConsequence.eventually(GivenWhenThen.seeThat(
   WebElementQuestion.the(LoginForm.PASSWORD),
   WebElementStateMatchers.isEnabled())
).waitingForNoLongerThan(30).seconds())
```
there might be more ways..

`stale element reference: element is not attached to the page document` might happen when you are checking if the element is present and enabled, the page reloads and the element is is different one now. Is there any not DOM related way to make sure you can enter the passwort now? like waiting for the url to be `/login/password`

**Comment by SiKing** [comment]
Thank you for the answer!

The global timeouts I was already aware of.
I remember, when I was coding the "classic" Selenium Page Objects some advice about not mixing implicit end explicit waits. How does this work in Serenity, and specifically the Screenplay pattern. If I set `webdriver.timeouts.implicitlywait` and then in my code I also use something like `.waitingForNoLongerThan(30).seconds()`, do they interact in a "safe" way?

Kinda unrelated: I noticed there are a lot of links (and package namespaces in the code) to thucydides and often duplicated to serenity-bdd. Is serenity-bdd the newer / **preferred** one?

Your first code sample - perform Wait - gives me: "The method until(Question, Matcher) in the type Wait is not applicable for the arguments (Target, Matcher<WebElementState>)".

Your second code sample works. So I guess I will go with that for now.
